File: server/routes/compliance.js
/**
 * GDPR / CCPA Compliance Routes
 * ================================
 * Data privacy endpoints for regulatory compliance.
 *
 * Routes:
 *   GET    /compliance/consent          — Get consent status
 *   POST   /compliance/consent          — Record consent
 *   POST   /compliance/export           — Request data export
 *   GET    /compliance/export/:id       — Check export status
 *   POST   /compliance/delete           — Request data deletion
 *   GET    /compliance/privacy-policy   — Privacy policy metadata
 */

const express = require('express');
const db = require('../models/db');
const { v4: uuidv4 } = require('uuid');

const router = express.Router();

// =========================================
// Get Consent Status
// =========================================

router.get('/consent', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;

        const result = await db.query(
            'SELECT gdpr_consent, gdpr_consent_at FROM players WHERE id = $1 AND tenant_id = $2',
            [playerId, tenantId]
        );

        if (result.rows.length === 0) {
            return res.status(404).json({ error: 'Player not found' });
        }

        res.json({
            consent: result.rows[0].gdpr_consent,
            consentAt: result.rows[0].gdpr_consent_at
        });
    } catch (err) {
        next(err);
    }
});

// =========================================
// Record Consent
// =========================================

router.post('/consent', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;
        const { consent } = req.body;

        if (typeof consent !== 'boolean') {
            return res.status(400).json({ error: 'consent must be a boolean' });
        }

        await db.query(`
            UPDATE players SET gdpr_consent = $1, gdpr_consent_at = NOW()
            WHERE id = $2 AND tenant_id = $3
        `, [consent, playerId, tenantId]);

        res.json({ message: 'Consent recorded', consent });
    } catch (err) {
        next(err);
    }
});

// =========================================
// Request Data Export (GDPR Article 20)
// =========================================

router.post('/export', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;

        // Check for existing pending export
        const existing = await db.query(`
            SELECT id, status, created_at FROM gdpr_requests
            WHERE player_id = $1 AND tenant_id = $2 AND request_type = 'export'
                AND status IN ('pending', 'processing')
            ORDER BY created_at DESC LIMIT 1
        `, [playerId, tenantId]);

        if (existing.rows.length > 0) {
            return res.status(409).json({
                error: 'Export request already pending',
                requestId: existing.rows[0].id,
                status: existing.rows[0].status
            });
        }

        // Create export request
        const result = await db.query(`
            INSERT INTO gdpr_requests (tenant_id, player_id, request_type, status)
            VALUES ($1, $2, 'export', 'pending')
            RETURNING id, status, created_at
        `, [tenantId, playerId]);

        // Trigger async export (in production, this would be a job queue)
        _processExport(result.rows[0].id, playerId, tenantId).catch(err => {
            console.error('GDPR export failed:', err);
        });

        res.status(202).json({
            message: 'Data export request received. You will be notified when ready.',
            request: result.rows[0]
        });
    } catch (err) {
        next(err);
    }
});

// =========================================
// Check Export Status
// =========================================

router.get('/export/:id', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;
        const requestId = req.params.id;

        const result = await db.query(`
            SELECT id, status, download_url, completed_at, expires_at, created_at
            FROM gdpr_requests
            WHERE id = $1 AND player_id = $2 AND tenant_id = $3 AND request_type = 'export'
        `, [requestId, playerId, tenantId]);

        if (result.rows.length === 0) {
            return res.status(404).json({ error: 'Export request not found' });
        }

        res.json({ request: result.rows[0] });
    } catch (err) {
        next(err);
    }
});

// =========================================
// Request Data Deletion (GDPR Article 17)
// =========================================

router.post('/delete', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;
        const { confirmation } = req.body;

        if (confirmation !== 'DELETE_MY_DATA') {
            return res.status(400).json({
                error: 'Confirmation required. Send { "confirmation": "DELETE_MY_DATA" }'
            });
        }

        // Check for existing pending deletion
        const existing = await db.query(`
            SELECT id FROM gdpr_requests
            WHERE player_id = $1 AND tenant_id = $2 AND request_type = 'delete'
                AND status IN ('pending', 'processing')
        `, [playerId, tenantId]);

        if (existing.rows.length > 0) {
            return res.status(409).json({ error: 'Deletion request already pending' });
        }

        // Mark player for deletion (30-day grace period)
        await db.query(`
            UPDATE players SET data_deletion_requested_at = NOW()
            WHERE id = $1 AND tenant_id = $2
        `, [playerId, tenantId]);

        // Create deletion request
        const result = await db.query(`
            INSERT INTO gdpr_requests (tenant_id, player_id, request_type, status)
            VALUES ($1, $2, 'delete', 'pending')
            RETURNING id, status, created_at
        `, [tenantId, playerId]);

        res.status(202).json({
            message: 'Data deletion request received. Your data will be deleted within 30 days. You can cancel by contacting support.',
            request: result.rows[0]
        });
    } catch (err) {
        next(err);
    }
});

// =========================================
// Privacy Policy Metadata
// =========================================

router.get('/privacy-policy', (req, res) => {
    res.json({
        version: '1.0',
        lastUpdated: '2026-01-01',
        dataCollected: [
            { category: 'Account', items: ['email', 'display name', 'avatar'] },
            { category: 'Gameplay', items: ['scores', 'progress', 'match history'] },
            { category: 'Social', items: ['friend list', 'chat messages (in-game only)'] },
            { category: 'Technical', items: ['IP address (for matchmaking)', 'device type', 'connection quality'] }
        ],
        dataRetention: {
            accountData: 'Until account deletion',
            gameplayData: 'Until account deletion',
            chatMessages: '30 days',
            serverLogs: '90 days',
            anonymizedAnalytics: 'Indefinite'
        },
        thirdParties: [
            { name: 'Stripe', purpose: 'Payment processing' },
            { name: 'Google', purpose: 'SSO Authentication (if used)' },
            { name: 'Apple', purpose: 'SSO Authentication (if used)' }
        ],
        rights: [
            'Right to access your data (data export)',
            'Right to rectification (update profile)',
            'Right to erasure (delete account)',
            'Right to data portability (JSON export)',
            'Right to object to processing'
        ],
        contactEmail: 'privacy@minigames.cool'
    });
});

// =========================================
// Process Data Export (async)
// =========================================

async function _processExport(requestId, playerId, tenantId) {
    try {
        await db.query(
            "UPDATE gdpr_requests SET status = 'processing' WHERE id = $1",
            [requestId]
        );

        // Collect all player data
        const [profile, gameProgress, scores, matches, friends, transactions, inventory] = await Promise.all([
            db.query('SELECT id, display_name, email, avatar_character, region, locale, created_at FROM players WHERE id = $1 AND tenant_id = $2', [playerId, tenantId]),
            db.query('SELECT game_id, high_score, stars, level, play_count, updated_at FROM game_progress WHERE player_id = $1 AND tenant_id = $2', [playerId, tenantId]),
            db.query('SELECT game_id, score, created_at FROM score_history WHERE player_id = $1 AND tenant_id = $2 ORDER BY created_at DESC LIMIT 1000', [playerId, tenantId]),
            db.query('SELECT match_id, player_index, score, is_winner, placement, created_at FROM multiplayer_match_players WHERE player_id = $1 AND tenant_id = $2 ORDER BY created_at DESC LIMIT 500', [playerId, tenantId]),
            db.query(`SELECT p.display_name, f.status, f.created_at FROM friendships f JOIN players p ON (CASE WHEN f.player_id = $1 THEN p.id = f.friend_id ELSE p.id = f.player_id END) WHERE f.tenant_id = $2 AND (f.player_id = $1 OR f.friend_id = $1)`, [playerId, tenantId]),
            db.query('SELECT currency_type, amount, tx_type, source, created_at FROM economy_transactions WHERE player_id = $1 AND tenant_id = $2 ORDER BY created_at DESC LIMIT 500', [playerId, tenantId]),
            db.query('SELECT si.name, si.item_type, pi.acquired_at, pi.source FROM player_inventory pi JOIN store_items si ON si.id = pi.item_id WHERE pi.player_id = $1 AND pi.tenant_id = $2', [playerId, tenantId])
        ]);

        const exportData = {
            exportedAt: new Date().toISOString(),
            profile: profile.rows[0] || {},
            gameProgress: gameProgress.rows,
            scoreHistory: scores.rows,
            multiplayerMatches: matches.rows,
            friends: friends.rows,
            transactions: transactions.rows,
            inventory: inventory.rows
        };

        // In production: upload to S3/GCS and return a signed URL
        // For now: store as JSON reference
        const downloadUrl = `/api/v1/compliance/export/${requestId}/download`;

        await db.query(`
            UPDATE gdpr_requests
            SET status = 'completed', download_url = $1, completed_at = NOW(),
                expires_at = NOW() + INTERVAL '7 days'
            WHERE id = $2
        `, [downloadUrl, requestId]);

    } catch (err) {
        await db.query(
            "UPDATE gdpr_requests SET status = 'failed' WHERE id = $1",
            [requestId]
        );
        throw err;
    }
}

module.exports = router;

File: server-rs/src/routes/compliance.rs
use axum::{
    extract::{Path, State},
    Json,
};
use serde_json::{json, Value};
use uuid::Uuid;

use crate::error::{AppError, AppResult};
use crate::middleware::auth::AuthPlayer;
use crate::middleware::tenant::TenantId;
use crate::models::compliance::*;
use crate::AppState;

pub async fn get_consent(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
) -> AppResult<Json<Value>> {
    let row: Option<(Option<bool>, Option<chrono::DateTime<chrono::Utc>>)> = sqlx::query_as(
        "SELECT gdpr_consent, gdpr_consent_at FROM players WHERE id = $1 AND tenant_id = $2",
    )
    .bind(player.id)
    .bind(&tenant.0 .0)
    .fetch_optional(&state.db)
    .await?;

    match row {
        Some((consent, at)) => Ok(Json(json!({"consent": consent.unwrap_or(false), "consentAt": at}))),
        None => Ok(Json(json!({"consent": false}))),
    }
}

pub async fn record_consent(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
    Json(body): Json<ConsentRequest>,
) -> AppResult<Json<Value>> {
    sqlx::query(
        "UPDATE players SET gdpr_consent = $1, gdpr_consent_at = NOW() WHERE id = $2 AND tenant_id = $3",
    )
    .bind(body.consent)
    .bind(player.id)
    .bind(&tenant.0 .0)
    .execute(&state.db)
    .await?;

    Ok(Json(json!({"success": true})))
}

pub async fn request_export(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
) -> AppResult<Json<Value>> {
    let tid = &tenant.0 .0;

    let req_id: Uuid = sqlx::query_scalar(
        r#"INSERT INTO gdpr_requests (id, tenant_id, player_id, request_type, status, created_at)
        VALUES (gen_random_uuid(), $1, $2, 'export', 'pending', NOW())
        RETURNING id"#,
    )
    .bind(tid)
    .bind(player.id)
    .fetch_one(&state.db)
    .await?;

    // Process export inline (in production this would be async)
    let pid = player.id;
    let db = state.db.clone();
    let tenant_id = tid.clone();
    tokio::spawn(async move {
        let _ = process_export(&db, pid, &tenant_id, req_id).await;
    });

    Ok(Json(json!({"requestId": req_id, "status": "pending"})))
}

async fn process_export(
    db: &sqlx::PgPool,
    player_id: Uuid,
    tenant_id: &str,
    request_id: Uuid,
) -> AppResult<()> {
    // Collect player data
    let profile: Option<Value> = sqlx::query_scalar(
        "SELECT row_to_json(p) FROM (SELECT id, email, display_name, avatar_character, total_score, games_played, created_at FROM players WHERE id = $1 AND tenant_id = $2) p",
    )
    .bind(player_id)
    .bind(tenant_id)
    .fetch_optional(db)
    .await?;

    let progress: Vec<Value> = sqlx::query_scalar(
        "SELECT row_to_json(gp) FROM (SELECT game_id, high_score, stars, play_count FROM game_progress WHERE player_id = $1 AND tenant_id = $2) gp",
    )
    .bind(player_id)
    .bind(tenant_id)
    .fetch_all(db)
    .await?;

    let scores: Vec<Value> = sqlx::query_scalar(
        "SELECT row_to_json(sh) FROM (SELECT game_id, score, level, created_at FROM score_history WHERE player_id = $1 AND tenant_id = $2 ORDER BY created_at DESC LIMIT 1000) sh",
    )
    .bind(player_id)
    .bind(tenant_id)
    .fetch_all(db)
    .await?;

    let export_data = json!({
        "profile": profile,
        "gameProgress": progress,
        "scoreHistory": scores,
        "exportedAt": chrono::Utc::now(),
    });

    let download_url = format!("data:application/json;base64,{}", base64_encode(&export_data.to_string()));
    let expires = chrono::Utc::now() + chrono::Duration::days(7);

    sqlx::query(
        "UPDATE gdpr_requests SET status = 'completed', download_url = $1, completed_at = NOW(), expires_at = $2 WHERE id = $3",
    )
    .bind(&download_url)
    .bind(expires)
    .bind(request_id)
    .execute(db)
    .await?;

    Ok(())
}

fn base64_encode(s: &str) -> String {
    use base64::Engine;
    base64::engine::general_purpose::STANDARD.encode(s.as_bytes())
}

pub async fn get_export_status(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
    Path(id): Path<String>,
) -> AppResult<Json<Value>> {
    let req_id = Uuid::parse_str(&id)
        .map_err(|_| AppError::BadRequest("Invalid request ID".into()))?;

    let row: Option<(String, Option<String>, Option<chrono::DateTime<chrono::Utc>>, Option<chrono::DateTime<chrono::Utc>>)> = sqlx::query_as(
        "SELECT status, download_url, completed_at, expires_at FROM gdpr_requests WHERE id = $1 AND player_id = $2 AND tenant_id = $3 AND request_type = 'export'",
    )
    .bind(req_id)
    .bind(player.id)
    .bind(&tenant.0 .0)
    .fetch_optional(&state.db)
    .await?;

    match row {
        Some((status, url, completed, expires)) => Ok(Json(json!({
            "status": status, "downloadUrl": url, "completedAt": completed, "expiresAt": expires,
        }))),
        None => Err(AppError::NotFound("Export request not found".into())),
    }
}

pub async fn request_deletion(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
    Json(body): Json<DeleteRequest>,
) -> AppResult<Json<Value>> {
    if body.confirmation != "DELETE_MY_DATA" {
        return Err(AppError::BadRequest(
            "Must confirm with 'DELETE_MY_DATA'".into(),
        ));
    }

    let tid = &tenant.0 .0;

    sqlx::query(
        "UPDATE players SET data_deletion_requested_at = NOW() WHERE id = $1 AND tenant_id = $2",
    )
    .bind(player.id)
    .bind(tid)
    .execute(&state.db)
    .await?;

    sqlx::query(
        "INSERT INTO gdpr_requests (id, tenant_id, player_id, request_type, status, created_at) VALUES (gen_random_uuid(), $1, $2, 'delete', 'pending', NOW())",
    )
    .bind(tid)
    .bind(player.id)
    .execute(&state.db)
    .await?;

    Ok(Json(json!({
        "success": true,
        "message": "Deletion scheduled. Your data will be deleted within 30 days.",
    })))
}

pub async fn privacy_policy() -> Json<Value> {
    Json(json!({
        "version": "1.0",
        "lastUpdated": "2025-01-01",
        "dataCollected": ["email", "display name", "game progress", "scores", "purchase history"],
        "retention": "Data retained while account active. Deleted 30 days after deletion request.",
        "rights": ["access", "rectification", "erasure", "portability", "restriction"],
        "contact": "privacy@minigames.cool",
    }))
}
